subprojects/inject-generics/src/test/java/org/mockito/internal/util/reflection/generic/GenericTypeMatchTest.java

```diff
@@ -0,0 +1,421 @@
+/*
+ * Copyright (c) 2016 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockito.internal.util.reflection.generic;
+
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
+
+import java.lang.reflect.Field;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@SuppressWarnings("unused")
+public class GenericTypeMatchTest {
+
+    public static final String STATIC_CLASS_PREFIX = GenericTypeMatchTest.class.getName() + "$";
+
+    @Nested
+    public class SimpleTypesTest {
+
+        private Integer integer;
+        private Number number;
+        private String string;
+
+        @ParameterizedTest
+        @CsvSource({
+            "integer  ,integer  ,true",
+            "integer  ,number   ,true",
+            "number   ,integer  ,false",
+            "integer  ,string   ,false",
+            "string   ,integer  ,false"
+        })
+        public void testSimpleTypes(String sourceFieldName, String targetFieldName, boolean matches) throws NoSuchFieldException {
+            Field sourceField = SimpleTypesTest.class.getDeclaredField(sourceFieldName);
+            Field targetField = SimpleTypesTest.class.getDeclaredField(targetFieldName);
+            GenericTypeMatch sourceMatch = GenericTypeMatch.ofField(sourceField);
+            GenericTypeMatch targetMatch = GenericTypeMatch.ofField(targetField);
+            assertEquals(matches, targetMatch.matches(sourceMatch));
+        }
+
+    }
+
+    @Nested
+    public class ParameterizedTypesTest {
+
+        private List<Integer> integerList;
+        private Collection<Integer> integerCollection;
+        private List<String> stringList;
+        private List<?> wildcardList;
+        private List<? extends Integer> wildcardListInteger;
+        private Map<String, Integer> mapStringInteger;
+        private Map<?, ?> wildcardMap;
+        private Map<? extends String, ? extends Number> wildcardMapStringNumber;
+
+        @ParameterizedTest
+        @CsvSource({
+            "integerList          ,integerList          ,true",
+            "integerList          ,integerCollection    ,true",
+            "integerCollection    ,integerList          ,false",
+            "integerList          ,stringList           ,false",
+            "stringList           ,integerList          ,false",
+
+            "stringList           ,wildcardList         ,true",
```

Comment:
Nit: add a case for `"integerList           ,wildcardListInteger         ,true",`